pkg/metrics/metrics.go

```diff
@@ -0,0 +1,195 @@
+		"device_free",
+		"Amount of Free space available on the device",
+		[]string{"cluster", "hostname", "device"},
+	)
+
+	deviceUsed = promDesc(
+		"device_used",
+		"Amount of space used on the device",
+		[]string{"cluster", "hostname", "device"},
+	)
+
+	brickCount = promDesc(
+		"device_brick_count",
+		"Number of bricks on device",
+		[]string{"cluster", "hostname", "device"},
+	)
+)
+
+func promDesc(name, help string, variableLabels []string) *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", name),
+		help,
+		variableLabels,
+		nil,
+	)
+}
+
+// Describe all the metrics exported by Heketi exporter. It implements prometheus.Collector.
+func (m *Metrics) Describe(ch chan<- *prometheus.Desc) {
+	ch <- clusterCount
+	ch <- volumesCount
+	ch <- nodesCount
+	ch <- deviceCount
+	ch <- deviceSize
+	ch <- deviceFree
+	ch <- deviceUsed
+	ch <- brickCount
+
+}
+
+// Collect metrics from heketi app
+func (m *Metrics) Collect(ch chan<- prometheus.Metric) {
+	topinfo, err := m.app.TopologyInfo()
+
+	upVal := 0.0
+	if err == nil {
+		upVal = 1.0
+	}
+
+	ch <- prometheus.MustNewConstMetric(up, prometheus.GaugeValue, upVal)
+
+	//Do not collect further metrics if heketi is down
+	if err != nil {
+		log.Println("Can't collect toplogy info for metrics: " + err.Error())
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		clusterCount,
+		prometheus.GaugeValue,
+		float64(len(topinfo.ClusterList)),
+	)
+	for _, cluster := range topinfo.ClusterList {
+		ch <- prometheus.MustNewConstMetric(
+			volumesCount,
+			prometheus.GaugeValue,
+			float64(len(cluster.Volumes)),
+			cluster.Id,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			nodesCount,
+			prometheus.GaugeValue,
+			float64(len(cluster.Nodes)),
+			cluster.Id,
+		)
+		for _, node := range cluster.Nodes {
+			ch <- prometheus.MustNewConstMetric(
+				deviceCount,
+				prometheus.GaugeValue,
+				float64(len(node.DevicesInfo)),
+				cluster.Id,
+				node.Hostnames.Manage[0],
+			)
+			for _, device := range node.DevicesInfo {
+				ch <- prometheus.MustNewConstMetric(
+					deviceSize,
+					prometheus.GaugeValue,
+					float64(device.Storage.Total),
+					cluster.Id,
+					node.Hostnames.Manage[0],
+					device.Name,
+				)
+				ch <- prometheus.MustNewConstMetric(
+					deviceFree,
+					prometheus.GaugeValue,
+					float64(device.Storage.Free),
+					cluster.Id,
+					node.Hostnames.Manage[0],
+					device.Name,
+				)
+				ch <- prometheus.MustNewConstMetric(
+					deviceUsed, prometheus.GaugeValue,
+					float64(device.Storage.Used),
+					cluster.Id,
+					node.Hostnames.Manage[0],
+					device.Name,
+				)
+				ch <- prometheus.MustNewConstMetric(
+					brickCount,
+					prometheus.GaugeValue,
+					float64(len(device.Bricks)),
+					cluster.Id,
+					node.Hostnames.Manage[0],
+					device.Name,
+				)
+			}
+		}
+	}
+}
+
+func NewMetricsHandler(app apps.Application) http.HandlerFunc {
+	m := &Metrics{
+		app: app,
+	}
+	prometheus.MustRegister(m)
```

Comment:
What does this do? Are we always requiring some prometheus service with this change?

Reply:
Must Register starts the gather loop for metrics. It's not really a service is the traditional sense. Here are the docs for it: https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L89

Reply:
ok